engines/cufile.c

```diff
@@ -0,0 +1,586 @@
+								cufileop_status_error(-sz));
+					}
+				} else {
+					sz = pwrite(io_u->file->fd,
+								((char*) io_u->xfer_buf) + xfered,
+								remaining, io_offset + xfered);
+					if (sz < 0) {
+						io_u->error = errno;
+						log_err("pwrite: %d\n", errno);
+					}
+				}
+			} else {
+				/* shouldn't happen */
+				assert(0);
+				log_err("not DDIR_READ or DDIR_WRITE: %d\n", io_u->ddir);
+				io_u->error = -1;
+				break;
+			}
+
+			if (io_u->error != 0) {
+				break;
+			}
+
+			remaining -= sz;
+			xfered += sz;
+
+			if (remaining != 0) {
+				log_info("Incomplete %s: %ld bytes remaining\n",
+						 io_u->ddir == DDIR_READ? "read" : "write", remaining);
+			}
+
+		}
+
+		if (io_u->error != 0) {
+			break;
+		}
+
+		if (io_u->ddir == DDIR_READ) {
+			rc = fio_cufile_post_read(td, o, io_u, gpu_offset);
+		}
+		break;
+
+	default:
+		io_u->error = EINVAL;
+		break;
+	}
+
+	if (io_u->error != 0) {
+		log_err("IO failed\n");
+		td_verror(td, io_u->error, "xfer");
+	}
+
+	return FIO_Q_COMPLETED;
+}
+
+static int fio_cufile_open_file(struct thread_data *td, struct fio_file *f)
+{
+	struct cufile_options *o = td->eo;
+	struct fio_cufile_data *fcd = NULL;
+	int rc;
+	CUfileError_t status;
+
+	rc = generic_open_file(td, f);
+	if (rc) {
+		return rc;
+	}
+
+	if (!o->use_posixio) {
+		fcd = calloc(1, sizeof(*fcd));
+		if (!fcd) {
+			rc = ENOMEM;
+			goto exit_err;
+		}
+
+		fcd->cf_descr.handle.fd = f->fd;
+		fcd->cf_descr.type = CU_FILE_HANDLE_TYPE_OPAQUE_FD;
+		status = cuFileHandleRegister(&fcd->cf_handle, &fcd->cf_descr);
+		if (status.err != CU_FILE_SUCCESS) {
+			log_err("cufile register error: %d:%s\n", status.err,
+					cuFileGetErrorString(status));
+			rc = EINVAL;
+			goto exit_err;
+		}
+	}
+
+	FILE_SET_ENG_DATA(f, fcd);
+	return 0;
+
+exit_err:
+	if (fcd) {
+		free(fcd);
+		fcd = NULL;
+	}
+	if (f) {
+		int rc2 = generic_close_file(td, f);
+		if (rc2)
+			log_err("generic_close_file %s:%d : %d\n", __func__, __LINE__, rc2);
+	}
+	return rc;
+}
+
+static int fio_cufile_close_file(struct thread_data *td, struct fio_file *f)
+{
+	struct fio_cufile_data *fcd = FILE_ENG_DATA(f);
+	int rc;
+
+	if (fcd != NULL) {
+		cuFileHandleDeregister(fcd->cf_handle);
+		FILE_SET_ENG_DATA(f, NULL);
+		free(fcd);
+	}
+
+	rc = generic_close_file(td, f);
+
+	return rc;
+}
+
+static int fio_cufile_iomem_alloc(struct thread_data *td, size_t total_mem)
+{
+	struct cufile_options *o = td->eo;
+	int rc;
+	CUfileError_t status;
+
+	o->total_mem = total_mem;
+	o->logged = 0;
+	td->orig_buffer = calloc(1, total_mem);
+	if (!td->orig_buffer) {
+		log_err("calloc failed: %d\n", errno);
+		goto exit_error;
+	}
+
+	if (o->use_posixio) {
+		o->junk_buf = calloc(1, total_mem);
+		if (o->junk_buf == NULL) {
+			goto exit_error;
+		}
+	} else {
+		o->junk_buf = NULL;
+	}
+
+	dprint(FD_MEM, "Alloc %zu for GPU %d\n", total_mem, o->my_gpu_id);
+	check_cudaruntimecall(cudaMalloc(&o->cu_mem_ptr, total_mem), rc);
+	if (rc != 0) {
+		goto exit_error;
+	}
```

Comment:
House style is that single line conditional bodies without an else don't get braces.

Reply:
OK, will do.